libnetwork/driver_plugin.py

```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
 # The API calls below are documented at
 # https://github.com/docker/libnetwork/blob/master/docs/remote.md
 
+# <-- Plugin activation, we activate both libnetwork and ipam plugins -->
 
 @app.route('/Plugin.Activate', methods=['POST'])
 def activate():
-    json_response = {"Implements": ["NetworkDriver"]}
+    json_response = {"Implements": ["NetworkDriver", "IpamDriver"]}
     app.logger.debug("Activate response JSON=%s", json_response)
     return jsonify(json_response)
 
+# <-- IPAM plugin API -->
+
+@app.route('/IpamDriver.GetDefaultAddressSpaces', methods=['POST'])
+def get_default_address_spaces():
+    json_data = request.get_json(force=True)
+    app.logger.debug("GetDefaultAddressSpace JSON=%s", json_data)
+    json_response = {
+        "LocalDefaultAddressSpace": LOCAL_CIDR_ADDR_SP,
+        "GlobalDefaultAddressSpace": GLOBAL_CIDR_ADDR_SP
+    }
+    app.logger.debug("GetDefaultAddressSpace response JSON=%s", json_response)
+    return jsonify(json_response)
+
+
+@app.route('/IpamDriver.RequestPool', methods=['POST'])
+def request_pool():
+    json_data = request.get_json(force=True)
+    app.logger.debug("RequestPool JSON=%s", json_data)
+    address_space = json_data["AddressSpace"]
+    pool = json_data["Pool"]
+    sub_pool = json_data.get("SubPool")
+    options = json_data["Options"]
+    v6 = json_data["V6"]
+
+    # Calico IPAM does not allow you to request Pool or SubPool.
+    assert not pool and not sub_pool, \
+        "Calico IPAM does not support pool configuration on 'docker create " \
+        "network'.  Calico IP Pools should be configured first and IP " \
+        "assignment is from those pre-configured pools."
+
+    # We use static pool ID and CIDR - these are arbitrary since we do not
+    # actually create a pool here.  It is assumed the user has configured
+    # all of the required Calico pools.
+    if v6:
+        pool_id = POOL_ID_V6
+        pool_cidr = POOL_CIDR_STR_V6
+        gway_cidr = GWAY_CIDR_STR_V6
+    else:
+        pool_id = POOL_ID_V4
+        pool_cidr = POOL_CIDR_STR_V4
+        gway_cidr = GWAY_CIDR_STR_V4
+
+    # The meta data includes a dummy gateway address.  This prevents libnetwork
+    # from requesting a gateway address from the pool since for a Calico
+    # network we route to the host.
```

Comment:
...which means the route/gateway is set up per container (in join()) rather than per network.